gpflow/utilities/utilities.py

```diff
@@ -44,6 +46,21 @@ def to_default_float(x):
     return tf.cast(x, dtype=default_float())
 
 
+class ShapeError(AssertionError):
+    pass
+
+
+def assert_shapes(shapes, *args, **kwargs):
+    """
+    Wrapper around tf.debugging.assert_shapes() to replace ValueError with ShapeError
+    """
+    if __debug__:
+        try:
+            tf.debugging.assert_shapes(shapes, *args, **kwargs)
+        except ValueError as e:
```

Comment:
you mentioned it could also raise InvalidArgumentError
```suggestion
        except (tf.errors.InvalidArgumentError, ValueError) as e:
```

Reply:
Yes, but that won't actually work as the try:except: is parsed at trace-time, whereas the InvalidArgumentError would only be raised at graph-run-time, so it will still be an InvalidArgumentError (I just tried it out locally).

Reply:
i'm starting to question the sanity of something like `tf.function` in an interpreted language

Reply:
so i guess the problem is that try except doesn't work as expected in `tf.function`. OK, if we can't make this consistent with and without `tf.function`, maybe we should abandon the custom error.

How about we just use `tf.debugging.assert_shapes` as is (delete the custom `assert_shapes` function), and apply `if __debug__` where checks are for code consistency, then at least users can just turn those ones off and catch the `InvalidArgumentError` or `ValueError` as normal from outside the `tf.function`.

Sorry we've had to go round the houses. I'm confident it was worth looking into, but alas I think we're limited here by python/tensorflow functionality

Reply:
I don't think the `if __debug__:` all around our code are worth it. I'm happy to kick out the wrapper function and replace with tf.debugging.assert_shapes again though.

Reply:
ok. i think it could be a _bit_ of a problem for users, but let's just do it and get the PR in